File: python/src/fluke_3540/events.py
"""Event detection for Fluke 3540 FC sessions.

Detects outages, voltage dips/swells, high-current excursions, frequency
deviations, imbalance spikes, and large power steps. Stdlib only — no numpy.

Thresholds live in EVENT_RULES and follow IEEE 1159 / NEMA conventions.
See docs/EVENT_RULES.md for the rationale.
"""
from __future__ import annotations

import datetime as dt
import math
from dataclasses import dataclass
from statistics import median, mean, pstdev
from typing import Iterable, Sequence

from .parser import FIELDS, Record


# --- Field index lookup -------------------------------------------------------

_FIELD_INDEX = {f.name: f.index for f in FIELDS}


def _idx(name: str) -> int:
    try:
        return _FIELD_INDEX[name]
    except KeyError as e:
        raise KeyError(
            f"Field {name!r} missing from spec/field_map.json — events module "
            "depends on it"
        ) from e


# Voltage channels (per-phase L-N min/max/avg)
_V_LN_MIN = (_idx("V_LN_a_min_V"), _idx("V_LN_b_min_V"), _idx("V_LN_c_min_V"))
_V_LN_MAX = (_idx("V_LN_a_max_V"), _idx("V_LN_b_max_V"), _idx("V_LN_c_max_V"))
_V_LN_AVG = (_idx("V_LN_a_avg_V"), _idx("V_LN_b_avg_V"), _idx("V_LN_c_avg_V"))

# Current channels
_I_MAX = (_idx("I_a_max_A"), _idx("I_b_max_A"), _idx("I_c_max_A"))

# Frequency (use avg)
_FREQ_AVG = _idx("freq_avg_Hz")

# Total active power (avg over 1-sec window)
_P_TOTAL_AVG = _idx("P_total_avg_W")

_PHASES = ("a", "b", "c")


# --- Rule constants -----------------------------------------------------------

@dataclass(frozen=True)
class EventRules:
    """Tunable detection thresholds. Defaults follow plan + IEEE 1159 / NEMA."""
    outage_v_threshold: float = 50.0          # any phase L-N below this V is "outage"
    dip_pct_of_nominal: float = 0.90          # < 90% nominal = dip
    swell_pct_of_nominal: float = 1.10        # > 110% nominal = swell
    high_current_sigma: float = 2.0           # mean + 2σ on any phase
    freq_excursion_hz: float = 0.5            # |f - 60| > 0.5 Hz
    imbalance_pct_threshold: float = 2.5      # NEMA % imbalance threshold
    power_step_pct_of_mean: float = 0.50      # ΔP within 1 sec > 50% session mean |P|
    min_duration_secs: int = 1                # below this an event is ignored
    gap_tolerance_secs: int = 1               # merge runs separated by ≤ this gap
    nominal_freq_hz: float = 60.0


DEFAULT_RULES = EventRules()


@dataclass(frozen=True)
class Event:
    id: int
    kind: str             # outage | dip | swell | high_current | freq_excursion |
                          # imbalance_spike | power_step
    t_start: dt.datetime
    t_end: dt.datetime
    severity: float       # kind-specific magnitude (see comments by each detector)
    affected_phases: tuple[str, ...]


# --- Helpers ------------------------------------------------------------------

def _group_runs(mask: Sequence[bool], gap_tolerance: int) -> list[tuple[int, int]]:
    """Return contiguous [start, end] index ranges where mask is True.
    Runs separated by ≤ gap_tolerance False samples are merged.
    """
    runs: list[tuple[int, int]] = []
    in_run = False
    start = 0
    gap = 0
    for i, m in enumerate(mask):
        if m:
            if not in_run:
                in_run = True
                start = i
            gap = 0
        else:
            if in_run:
                gap += 1
                if gap > gap_tolerance:
                    runs.append((start, i - gap))
                    in_run = False
                    gap = 0
    if in_run:
        runs.append((start, len(mask) - 1 - max(0, gap)))
    return runs


def _infer_nominal_ln_v(v_avg_by_phase: list[list[float]],
                        outage_threshold: float) -> float:
    """Median of per-phase avg L-N voltages, with outage samples excluded."""
    pooled = [
        v
        for phase_vals in v_avg_by_phase
        for v in phase_vals
        if v > outage_threshold and not math.isnan(v)
    ]
    if not pooled:
        raise ValueError("No non-outage voltage samples — cannot infer nominal V_LN")
    return median(pooled)


# --- Main entry point ---------------------------------------------------------

def detect_events(
    records: Iterable[Record],
    nominal_ln_v: float | None = None,
    rules: EventRules = DEFAULT_RULES,
) -> list[Event]:
    """Scan a sequence of Records and return detected events in time order.

    If nominal_ln_v is None, it is auto-inferred as the median of per-phase L-N
    avg voltages with outage samples (<= outage_v_threshold) excluded.
    """
    recs = list(records)
    if not recs:
        return []

    # Column extraction
    times = [r.start for r in recs]
    end_times = [r.end for r in recs]
    v_min = [[r.floats[idx] for r in recs] for idx in _V_LN_MIN]
    v_max = [[r.floats[idx] for r in recs] for idx in _V_LN_MAX]
    v_avg = [[r.floats[idx] for r in recs] for idx in _V_LN_AVG]
    i_max = [[r.floats[idx] for r in recs] for idx in _I_MAX]
    freq = [r.floats[_FREQ_AVG] for r in recs]
    p_total = [r.floats[_P_TOTAL_AVG] for r in recs]

    if nominal_ln_v is None:
        nominal_ln_v = _infer_nominal_ln_v(v_avg, rules.outage_v_threshold)
    dip_threshold = nominal_ln_v * rules.dip_pct_of_nominal
    swell_threshold = nominal_ln_v * rules.swell_pct_of_nominal

    # not_outage[i] is True when ALL phases > outage_v_threshold (avg voltage)
    not_outage = [
        all(v_avg[p][i] > rules.outage_v_threshold for p in range(3))
        for i in range(len(recs))
    ]
    outage_mask = [not n for n in not_outage]

    events: list[Event] = []
    next_id = [0]

    def emit(kind: str, s: int, e: int, severity: float, phases: Iterable[str]):
        if (e - s + 1) < rules.min_duration_secs:
            return
        events.append(Event(
            id=next_id[0], kind=kind,
            t_start=times[s], t_end=end_times[e],
            severity=float(severity),
            affected_phases=tuple(phases),
        ))
        next_id[0] += 1

    # --- outage --------------------------------------------------------------
    # Severity: deepest min L-N voltage seen during the window (lower = worse).
    for s, e in _group_runs(outage_mask, rules.gap_tolerance_secs):
        lows = [min(v_min[p][s:e + 1]) for p in range(3)]
        emit("outage", s, e, severity=min(lows), phases=_PHASES)

    # --- dip ----------------------------------------------------------------
    # V_LN_min < dip_threshold on any phase, AND not currently an outage.
    # Severity: deepest dip below nominal as a fraction (lower = worse).
    dip_mask = [
        not_outage[i]
        and any(v_min[p][i] < dip_threshold for p in range(3))
        for i in range(len(recs))
    ]
    for s, e in _group_runs(dip_mask, rules.gap_tolerance_secs):
        dipped = []
        deepest = float("inf")
        for p in range(3):
            phase_min = min(v_min[p][s:e + 1])
            if phase_min < dip_threshold:
                dipped.append(_PHASES[p])
                deepest = min(deepest, phase_min)
        emit("dip", s, e, severity=deepest / nominal_ln_v, phases=dipped)

    # --- swell --------------------------------------------------------------
    # V_LN_max > swell_threshold on any phase, while not in outage.
    # Severity: highest swell above nominal as a fraction (higher = worse).
    swell_mask = [
        not_outage[i]
        and any(v_max[p][i] > swell_threshold for p in range(3))
        for i in range(len(recs))
    ]
    for s, e in _group_runs(swell_mask, rules.gap_tolerance_secs):
        swelled = []
        highest = float("-inf")
        for p in range(3):
            phase_max = max(v_max[p][s:e + 1])
            if phase_max > swell_threshold:
                swelled.append(_PHASES[p])
                highest = max(highest, phase_max)
        emit("swell", s, e, severity=highest / nominal_ln_v, phases=swelled)

    # --- high_current -------------------------------------------------------
    # Per-phase: i_max[p] > mean(i_max[p]) + N*stdev(i_max[p]).
    # Severity: peak A across all flagged phases.
    for p in range(3):
        # Mask outage samples — current spikes during outages are restoration
        # transients, classified under "outage" already.
        valid = [v for i, v in enumerate(i_max[p]) if not_outage[i]]
        if len(valid) < 2:
            continue
        mu = mean(valid)
        sigma = pstdev(valid) if len(valid) >= 2 else 0.0
        if sigma == 0.0:
            continue
        threshold = mu + rules.high_current_sigma * sigma
        mask = [
            not_outage[i] and i_max[p][i] > threshold
            for i in range(len(recs))
        ]
        for s, e in _group_runs(mask, rules.gap_tolerance_secs):
            peak = max(i_max[p][s:e + 1])
            emit("high_current", s, e, severity=peak, phases=(_PHASES[p],))

    # --- freq_excursion -----------------------------------------------------
    # Severity: signed deviation in Hz (negative = under-frequency).
    freq_mask = [
        not_outage[i] and abs(freq[i] - rules.nominal_freq_hz) > rules.freq_excursion_hz
        for i in range(len(recs))
    ]
    for s, e in _group_runs(freq_mask, rules.gap_tolerance_secs):
        window = freq[s:e + 1]
        # Take the most-extreme deviation in the window
        worst = max(window, key=lambda f: abs(f - rules.nominal_freq_hz))
        emit("freq_excursion", s, e,
             severity=worst - rules.nominal_freq_hz, phases=())

    # --- imbalance_spike ----------------------------------------------------
    # NEMA: max-min across phase avg L-N voltages > N% of mean.
    imbal_mask = []
    imbal_pct = []
    for i in range(len(recs)):
        if not not_outage[i]:
            imbal_mask.append(False)
            imbal_pct.append(0.0)
            continue
        vs = [v_avg[p][i] for p in range(3)]
        m = sum(vs) / 3.0
        if m <= rules.outage_v_threshold:
            imbal_mask.append(False)
            imbal_pct.append(0.0)
            continue
        pct = (max(vs) - min(vs)) / m * 100.0
        imbal_pct.append(pct)
        imbal_mask.append(pct > rules.imbalance_pct_threshold)
    for s, e in _group_runs(imbal_mask, rules.gap_tolerance_secs):
        peak_pct = max(imbal_pct[s:e + 1])
        emit("imbalance_spike", s, e, severity=peak_pct, phases=_PHASES)

    # --- power_step ---------------------------------------------------------
    # 1-sec ΔP_total exceeding N% of session mean |P_total| (outage-masked).
    # Severity: signed step magnitude in W (positive = sudden import, negative = export).
    p_valid = [p for i, p in enumerate(p_total) if not_outage[i]]
    if len(p_valid) >= 2:
        baseline = mean(abs(p) for p in p_valid)
        step_threshold = baseline * rules.power_step_pct_of_mean
        if step_threshold > 0:
            for i in range(1, len(recs)):
                if not (not_outage[i] and not_outage[i - 1]):
                    continue
                delta = p_total[i] - p_total[i - 1]
                if abs(delta) > step_threshold:
                    emit("power_step", i, i, severity=delta, phases=_PHASES)

    # Re-sort by time (high_current and others were grouped per phase, may
    # interleave) and re-issue sequential ids so the final list is canonical.
    events.sort(key=lambda ev: (ev.t_start, ev.kind))
    return [
        Event(id=i, kind=ev.kind, t_start=ev.t_start, t_end=ev.t_end,
              severity=ev.severity, affected_phases=ev.affected_phases)
        for i, ev in enumerate(events)
    ]

File: web/events.js
// Event detection — JS port of python/src/fluke_3540/events.py.
// Mirrors the same thresholds, mask logic, and Event shape.

import { asColumnSource } from './column_source.js';

export const DEFAULT_RULES = Object.freeze({
  outage_v_threshold: 50.0,
  dip_pct_of_nominal: 0.90,
  swell_pct_of_nominal: 1.10,
  high_current_sigma: 2.0,
  freq_excursion_hz: 0.5,
  imbalance_pct_threshold: 2.5,
  power_step_pct_of_mean: 0.50,
  min_duration_secs: 1,
  gap_tolerance_secs: 1,
  nominal_freq_hz: 60.0,
});

const PHASES = ['a', 'b', 'c'];

// Valid EventRules keys for --rules-file overrides (Feature I). Mirrors the
// Python EventRules dataclass fields.
const RULE_KEYS = new Set(Object.keys(DEFAULT_RULES));

/**
 * Resolve per-asset EventRules from a parsed rules-file object (JSON form),
 * mirroring python rules_file.load_rules. Precedence: DEFAULT_RULES ->
 * file.defaults -> file.assets[assetName] (or file.assets.default). A flat
 * object with only rule keys is treated as defaults.
 *
 * @param {object} raw parsed rules object
 * @param {string|null} assetName
 * @returns {object} a rules object suitable for detectEvents({rules})
 */
export function rulesFromObject(raw, assetName = null) {
  if (!raw || typeof raw !== 'object') return { ...DEFAULT_RULES };
  let defaults;
  let assets;
  if ('defaults' in raw || 'assets' in raw) {
    defaults = raw.defaults || {};
    assets = raw.assets || {};
  } else {
    defaults = raw;
    assets = {};
  }
  const coerce = (over, where) => {
    const out = {};
    for (const [k, v] of Object.entries(over)) {
      if (!RULE_KEYS.has(k)) {
        throw new Error(`${where}: unknown EventRules key ${k}`);
      }
      out[k] = (k === 'min_duration_secs' || k === 'gap_tolerance_secs')
        ? Math.trunc(Number(v)) : Number(v);
    }
    return out;
  };
  const merged = { ...DEFAULT_RULES, ...coerce(defaults, 'defaults') };
  let assetOver = {};
  if (assetName && assets[assetName]) assetOver = assets[assetName];
  else if (assets.default) assetOver = assets.default;
  return { ...merged, ...coerce(assetOver, 'assets') };
}

function fieldIndex(spec, name) {
  const f = spec.fields.find((f) => f.name === name);
  if (!f) throw new Error(`spec is missing field ${name}`);
  return f.index;
}

function median(values) {
  if (values.length === 0) return NaN;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0
    ? (sorted[mid - 1] + sorted[mid]) / 2
    : sorted[mid];
}

function mean(values) {
  if (values.length === 0) return NaN;
  let s = 0;
  for (const v of values) s += v;
  return s / values.length;
}

function pstdev(values) {
  if (values.length < 2) return 0;
  const m = mean(values);
  let s = 0;
  for (const v of values) s += (v - m) * (v - m);
  return Math.sqrt(s / values.length);
}

function groupRuns(mask, gapTolerance) {
  const runs = [];
  let inRun = false;
  let start = 0;
  let gap = 0;
  for (let i = 0; i < mask.length; i++) {
    if (mask[i]) {
      if (!inRun) { inRun = true; start = i; }
      gap = 0;
    } else if (inRun) {
      gap++;
      if (gap > gapTolerance) {
        runs.push([start, i - gap]);
        inRun = false;
        gap = 0;
      }
    }
  }
  if (inRun) {
    runs.push([start, mask.length - 1 - Math.max(0, gap)]);
  }
  return runs;
}

function inferNominalLnV(vAvgByPhase, outageThreshold) {
  const pooled = [];
  for (const phaseVals of vAvgByPhase) {
    for (const v of phaseVals) {
      if (v > outageThreshold && Number.isFinite(v)) pooled.push(v);
    }
  }
  if (pooled.length === 0) {
    throw new Error('No non-outage voltage samples — cannot infer nominal V_LN');
  }
  return median(pooled);
}

/**
 * Detect events on an array of parsed Records OR a ColumnStore.
 * @param {Array<{index:number, startMs:number, endMs:number, floats:Float32Array}>|import('./column_store.js').ColumnStore} source
 * @param {object} spec - parsed field_map.json
 * @param {{nominalLnV?: number, rules?: object}} [opts]
 * @returns {Array<Event>}
 */
export function detectEvents(source, spec, opts = {}) {
  const rules = { ...DEFAULT_RULES, ...(opts.rules ?? {}) };
  let { nominalLnV = null } = opts;
  const src = asColumnSource(source, spec);
  if (src.length === 0) return [];

  const N = src.length;
  // Extract columns once into typed arrays for speed (no-copy for a store).
  const vMin = PHASES.map((ph) => src.column(`V_LN_${ph}_min_V`));
  const vMax = PHASES.map((ph) => src.column(`V_LN_${ph}_max_V`));
  const vAvg = PHASES.map((ph) => src.column(`V_LN_${ph}_avg_V`));
  const iMaxArr = PHASES.map((ph) => src.column(`I_${ph}_max_A`));
  const freqArr = src.column('freq_avg_Hz');
  const pTotal = src.column('P_total_avg_W');

  if (nominalLnV === null) {
    nominalLnV = inferNominalLnV(vAvg, rules.outage_v_threshold);
  }
  const dipThreshold = nominalLnV * rules.dip_pct_of_nominal;
  const swellThreshold = nominalLnV * rules.swell_pct_of_nominal;

  const notOutage = new Array(N);
  for (let i = 0; i < N; i++) {
    notOutage[i] =
      vAvg[0][i] > rules.outage_v_threshold &&
      vAvg[1][i] > rules.outage_v_threshold &&
      vAvg[2][i] > rules.outage_v_threshold;
  }
  const outageMask = notOutage.map((b) => !b);

  const events = [];
  const startMs = (i) => src.startMs(i);
  const endMs = (i) => src.endMs(i);

  const phaseChars = (phaseList) => phaseList.slice();

  function rangeMin(arr, s, e) {
    let m = arr[s];
    for (let i = s + 1; i <= e; i++) if (arr[i] < m) m = arr[i];
    return m;
  }
  function rangeMax(arr, s, e) {
    let m = arr[s];
    for (let i = s + 1; i <= e; i++) if (arr[i] > m) m = arr[i];
    return m;
  }

  // outage
  for (const [s, e] of groupRuns(outageMask, rules.gap_tolerance_secs)) {
    if (e - s + 1 < rules.min_duration_secs) continue;
    const lows = vMin.map((arr) => rangeMin(arr, s, e));
    events.push({
      kind: 'outage', tStartMs: startMs(s), tEndMs: endMs(e),
      severity: Math.min(...lows), affectedPhases: phaseChars(PHASES),
    });
  }

  // dip
  const dipMask = new Array(N);
  for (let i = 0; i < N; i++) {
    dipMask[i] = notOutage[i] && (vMin[0][i] < dipThreshold ||
      vMin[1][i] < dipThreshold || vMin[2][i] < dipThreshold);
  }
  for (const [s, e] of groupRuns(dipMask, rules.gap_tolerance_secs)) {
    if (e - s + 1 < rules.min_duration_secs) continue;
    let deepest = Infinity;
    const dipped = [];
    for (let p = 0; p < 3; p++) {
      const phaseMin = rangeMin(vMin[p], s, e);
      if (phaseMin < dipThreshold) {
        dipped.push(PHASES[p]);
        deepest = Math.min(deepest, phaseMin);
      }
    }
    events.push({
      kind: 'dip', tStartMs: startMs(s), tEndMs: endMs(e),
      severity: deepest / nominalLnV, affectedPhases: dipped,
    });
  }

  // swell
  const swellMask = new Array(N);
  for (let i = 0; i < N; i++) {
    swellMask[i] = notOutage[i] && (vMax[0][i] > swellThreshold ||
      vMax[1][i] > swellThreshold || vMax[2][i] > swellThreshold);
  }
  for (const [s, e] of groupRuns(swellMask, rules.gap_tolerance_secs)) {
    if (e - s + 1 < rules.min_duration_secs) continue;
    let highest = -Infinity;
    const swelled = [];
    for (let p = 0; p < 3; p++) {
      const phaseMax = rangeMax(vMax[p], s, e);
      if (phaseMax > swellThreshold) {
        swelled.push(PHASES[p]);
        highest = Math.max(highest, phaseMax);
      }
    }
    events.push({
      kind: 'swell', tStartMs: startMs(s), tEndMs: endMs(e),
      severity: highest / nominalLnV, affectedPhases: swelled,
    });
  }

  // high_current
  for (let p = 0; p < 3; p++) {
    const valid = [];
    for (let i = 0; i < N; i++) if (notOutage[i]) valid.push(iMaxArr[p][i]);
    if (valid.length < 2) continue;
    const mu = mean(valid);
    const sigma = pstdev(valid);
    if (sigma === 0) continue;
    const threshold = mu + rules.high_current_sigma * sigma;
    const mask = new Array(N);
    for (let i = 0; i < N; i++) {
      mask[i] = notOutage[i] && iMaxArr[p][i] > threshold;
    }
    for (const [s, e] of groupRuns(mask, rules.gap_tolerance_secs)) {
      if (e - s + 1 < rules.min_duration_secs) continue;
      const peak = rangeMax(iMaxArr[p], s, e);
      events.push({
        kind: 'high_current', tStartMs: startMs(s), tEndMs: endMs(e),
        severity: peak, affectedPhases: [PHASES[p]],
      });
    }
  }

  // freq_excursion
  const freqMask = new Array(N);
  for (let i = 0; i < N; i++) {
    freqMask[i] = notOutage[i] && Math.abs(freqArr[i] - rules.nominal_freq_hz) > rules.freq_excursion_hz;
  }
  for (const [s, e] of groupRuns(freqMask, rules.gap_tolerance_secs)) {
    if (e - s + 1 < rules.min_duration_secs) continue;
    let worst = freqArr[s];
    for (let i = s + 1; i <= e; i++) {
      if (Math.abs(freqArr[i] - rules.nominal_freq_hz) > Math.abs(worst - rules.nominal_freq_hz)) {
        worst = freqArr[i];
      }
    }
    events.push({
      kind: 'freq_excursion', tStartMs: startMs(s), tEndMs: endMs(e),
      severity: worst - rules.nominal_freq_hz, affectedPhases: [],
    });
  }

  // imbalance_spike
  const imbalMask = new Array(N);
  const imbalPct = new Float64Array(N);
  for (let i = 0; i < N; i++) {
    if (!notOutage[i]) { imbalMask[i] = false; continue; }
    const v0 = vAvg[0][i], v1 = vAvg[1][i], v2 = vAvg[2][i];
    const m = (v0 + v1 + v2) / 3.0;
    if (m <= rules.outage_v_threshold) { imbalMask[i] = false; continue; }
    const pct = (Math.max(v0, v1, v2) - Math.min(v0, v1, v2)) / m * 100.0;
    imbalPct[i] = pct;
    imbalMask[i] = pct > rules.imbalance_pct_threshold;
  }
  for (const [s, e] of groupRuns(imbalMask, rules.gap_tolerance_secs)) {
    if (e - s + 1 < rules.min_duration_secs) continue;
    let peakPct = imbalPct[s];
    for (let i = s + 1; i <= e; i++) if (imbalPct[i] > peakPct) peakPct = imbalPct[i];
    events.push({
      kind: 'imbalance_spike', tStartMs: startMs(s), tEndMs: endMs(e),
      severity: peakPct, affectedPhases: PHASES.slice(),
    });
  }

  // power_step
  const pValid = [];
  for (let i = 0; i < N; i++) if (notOutage[i]) pValid.push(Math.abs(pTotal[i]));
  if (pValid.length >= 2) {
    const baseline = mean(pValid);
    const stepThreshold = baseline * rules.power_step_pct_of_mean;
    if (stepThreshold > 0) {
      for (let i = 1; i < N; i++) {
        if (!(notOutage[i] && notOutage[i - 1])) continue;
        const delta = pTotal[i] - pTotal[i - 1];
        if (Math.abs(delta) > stepThreshold) {
          events.push({
            kind: 'power_step', tStartMs: startMs(i), tEndMs: endMs(i),
            severity: delta, affectedPhases: PHASES.slice(),
          });
        }
      }
    }
  }

  // Sort by (t_start, kind) and re-issue sequential ids
  events.sort((a, b) => a.tStartMs - b.tStartMs || a.kind.localeCompare(b.kind));
  return events.map((ev, i) => ({ id: i, ...ev }));
}
